client/executor/common/src/sandbox/wasmer_backend.rs

```diff
@@ -0,0 +1,430 @@
+				ReturnValue::Value(Value::I64(val)) => vec![wasmer::Val::I64(val)],
+				ReturnValue::Value(Value::F32(val)) => vec![wasmer::Val::F32(f32::from_bits(val))],
+				ReturnValue::Value(Value::F64(val)) => vec![wasmer::Val::F64(f64::from_bits(val))],
+
+				ReturnValue::Unit => vec![],
+			};
+
+			Ok(result)
+		})
+		.expect("SandboxContextStore is set when invoking sandboxed functions; qed")
+	})
+}
+
+pub fn new_memory(
+	context: &Backend,
+	initial: u32,
+	maximum: Option<u32>,
+) -> crate::error::Result<Memory> {
+	let ty = wasmer::MemoryType::new(initial, maximum, false);
+	let memory = Memory::Wasmer(MemoryWrapper::new(
+		wasmer::Memory::new(&context.store, ty).map_err(|_| Error::InvalidMemoryReference)?,
+	));
+
+	Ok(memory)
+}
+
+/// In order to enforce memory access protocol to the backend memory
+/// we wrap it with `RefCell` and encapsulate all memory operations.
+#[derive(Debug, Clone)]
+pub struct MemoryWrapper {
+	buffer: Rc<RefCell<wasmer::Memory>>,
+}
+
+impl MemoryWrapper {
+	/// Take ownership of the memory region and return a wrapper object
+	pub fn new(memory: wasmer::Memory) -> Self {
+		Self { buffer: Rc::new(RefCell::new(memory)) }
+	}
+
+	/// Returns linear memory of the wasm instance as a slice.
+	///
+	/// # Safety
+	///
+	/// Wasmer doesn't provide comprehensive documentation about the exact behavior of the data
+	/// pointer. If a dynamic style heap is used the base pointer of the heap can change. Since
+	/// growing, we cannot guarantee the lifetime of the returned slice reference.
+	unsafe fn memory_as_slice(memory: &wasmer::Memory) -> &[u8] {
+		let ptr = memory.data_ptr() as *const _;
+		let len: usize = memory.data_size().try_into().expect("data size should fit into usize");
+
+		if len == 0 {
+			&[]
+		} else {
+			core::slice::from_raw_parts(ptr, len)
+		}
+	}
+
+	/// Returns linear memory of the wasm instance as a slice.
+	///
+	/// # Safety
+	///
+	/// See `[memory_as_slice]`. In addition to those requirements, since a mutable reference is
+	/// returned it must be ensured that only one mutable and no shared references to memory
+	/// exists at the same time.
+	unsafe fn memory_as_slice_mut(memory: &mut wasmer::Memory) -> &mut [u8] {
+		let ptr = memory.data_ptr();
+		let len: usize = memory.data_size().try_into().expect("data size should fit into usize");
+
+		if len == 0 {
+			&mut []
+		} else {
+			core::slice::from_raw_parts_mut(ptr, len)
+		}
+	}
+
+	/// Clone the underlying memory object
+	///
+	/// # Safety
+	///
+	/// The sole purpose of `MemoryRef` is to protect the memory from uncontrolled
+	/// access. By returning the memory object "as is" we bypass all of the checks.
+	///
+	/// Intended to use only during module initialization.
+	///
+	/// # Panics
+	///
+	/// Will panic if `MemoryRef` is currently in use.
+	pub unsafe fn clone_inner(&mut self) -> wasmer::Memory {
+		// We take exclusive lock to ensure that we're the only one here
+		self.buffer.borrow_mut().clone()
+	}
```

Comment:
Do we need this?

Reply:
Aside of trivial ones I'd like to address the comments in a separate PR.

Reply:
It was needed before, because parts of the logic were residing in different modules, which is not the case now. Fixed.